libcache/cache.c

```diff
@@ -0,0 +1,460 @@
+			if (set->tags[j] == NULL) {
+				set->tags[j] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+		set->count++;
+	}
+	else {
+		temp = set->timestamps;
+		LIST_REMOVE(&set->timestamps, temp);
+
+		/* write-back */
+		if (IS_DIRTY(temp->flags)) {
+			written = (*writeCb)(offset, temp->data, sizeof(*temp->data), count);
+			free(temp->data);
+		}
+
+		for (i = 0; i < LIBCACHE_NUM_WAYS; ++i) {
+			if (IS_VALID(set->lines[i].flags) && (set->lines[i].tag == temp->tag)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+
+	switch (policy) {
+		case 0:
+			/* write-back */
+			SET_DIRTY(set->lines[i].flags);
+			written = count;
+			break;
+		case 1:
+			/* write-through */
+			written = (*writeCb)(offset, temp->data, sizeof(*temp->data), count);
+			break;
+		default:
+			break;
+	}
+
+	qsort(set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return written;
+}
+
+
+static void *cache_readFromSet(cacheset_t *set, uint64_t tag, uint64_t offset)
+{
+	unsigned char *buf = NULL;
+	cacheline_t key, *keyPtr = &key, **linePtr = NULL;
+
+	key.tag = tag;
+
+	linePtr = bsearch(&keyPtr, set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	if (linePtr != NULL) {
+		buf = (unsigned char *)(*linePtr)->data + offset;
+
+		LIST_REMOVE(&set->timestamps, *linePtr);
+		LIST_ADD(&set->timestamps, *linePtr);
+	}
+
+	return buf;
+}
+
+
+static uint64_t cache_compOffset(const cachectx_t *cache, const uint64_t addr)
+{
+	return addr & cache->offMask;
+}
+
+
+static uint64_t cache_compSet(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> cache->offBitNum) & cache->setMask;
+}
+
+
+static uint64_t cache_compTag(const cachectx_t *cache, const uint64_t addr)
+{
+	return (addr >> (cache->offBitNum + cache->setBitsNum)) & cache->tagMask;
+}
+
+
+ssize_t cache_write(cachectx_t *cache, const uint64_t addr, void *buffer, size_t count, int policy)
+{
+	ssize_t wrt = 0;
+	void *dataPtr = NULL;
+	unsigned char *buf = NULL, *dest = NULL, flags = 0;
+	uint64_t index = 0, offset = 0, tag = 0;
+	cacheline_t line;
+
+	index = cache_compSet(cache, addr);
+	offset = cache_compOffset(cache, addr);
+	tag = cache_compTag(cache, addr);
+
+	dataPtr = malloc(sizeof(unsigned char) * cache->lineSize);
+
+	dest = (unsigned char *)dataPtr + offset;
+	/* TODO: make proper writing */
+	if (count > cache->lineSize - offset) {
+		count = cache->lineSize - offset;
+	}
+	memcpy(dest, buffer, count);
+
+	SET_VALID(flags);
+
+	line.tag = tag;
+	line.data = dataPtr;
+	line.flags = flags;
+	line.offset = offset;
+
+	return cache_writeToSet(cache->writeCb, &cache->sets[index], &line, offset, count, policy);
+}
+
+
+static ssize_t cache_readOnMiss(cachectx_t *cache, const uint64_t addr, const uint64_t offset, void *buffer, size_t position, size_t count)
+{
+	ssize_t read = 0, written = 0;
+	void *temp = NULL;
+
+	temp = calloc(cache->lineSize, sizeof(unsigned char));
```

Comment:
You don't check the value returned by calloc()

Reply:
I'll remember that, thanks